### DISCUSSION TITLE
Does pattern and format require a default value?

### DISCUSSION THREAD

**Discussion by thousynth** [sinclairzx81/typebox#613] [Q&A]
I'm trying to setup validations for email and password like following:
```js
const UserDTO = t.Object({
  name: t.String({}),
  email: t.String({
    format: 'email',
  }),
  password: t.String({
    minLength: 6,
    pattern: '/^(?=.*[a-z])(?=.*[A-Z])(?=.*d)(?=.*[@$!%*?&])[A-Za-zd@$!%*?&]{8,}$/',
  }),
})
```
rest of the code for route
```ts

const userRoutes = new Elysia({ prefix: '/auth' }).post('/signup', addUser, {
  body: UserDTO,
  beforeHandle: [checkEmail, checkName],
})
```
but when I use this to validate my body in a request it throws an error that `ValueCreate.String: String types with formats must specify a default value` if I remove the format from `email` it throws the same error but for `pattern` used in `password`: `ValueCreate.String: String types with patterns must specify a default value`.

Debugger throws from here: which feels like an elysia file. 
<img width="697" alt="image" src="https://github.com/sinclairzx81/typebox/assets/41203791/974fd054-fe77-4c6f-86f1-3f38802b232e">

Do they require defaults? I didn't see them in the docs.
Is this coming from bun or is this from typebox?

**Comment by sinclairzx81** [comment]
@cheersmas Hi,

> Is this coming from bun or is this from typebox?

This is from TypeBox. The `Value.Create` function cannot automatically create values for string formats (as there could be countless formats). Because of this, you do need to specify a default value when using formats if you intend to use Value.Create...

```typescript
const T = Type.String({ format: 'uuid', default: '00000000-0000-0000-0000-000000000000' })
```
The default value helps ensure that TypeBox generates a correct value.
Hope this helps
S

**Reply by thousynth** [reply]
hey, I still a couple of questions:
1. Do we still have to define a default if we are just using it for validations and not to use `Value.create`? I'm under the impression that Elysia's schemas are just for validations as mentioned [here](https://elysiajs.com/concept/schema.html). Is it still calling Value.create under the hood?

2. Are these some default `formats` or we have to register our own formats using `FormatRegistry`? For example I'm assuming that just setting the format to 'email' would assume that the input now should just be an email without me adding a format explicitly: 
<img width="614" alt="image" src="https://github.com/sinclairzx81/typebox/assets/41203791/0dbda8c2-b8e9-4ae9-928b-9bf5ab947c4f">

**Comment by sinclairzx81** [comment]
@cheersmas 

---

> Do we still have to define a default if we are just using it for validations and not to use Value.create? I'm under the impression that Elysia's schemas are just for validations as mentioned [here](https://elysiajs.com/concept/schema.html). Is it still calling Value.create under the hood?

TypeBox doesn't use `default` during validation. Note that it does not modify the value being checked or fill in empty values with `default` values if they are missing during a validation pass. You can omit the `default` property if you're just using validation. 

TypeBox implements `default` inline with the following

https://json-schema.org/understanding-json-schema/reference/generic.html

> The default keyword specifies a default value. _This value is not used to fill in missing values during the validation process_. Non-validation tools such as documentation generators or form generators may use this value to give hints to users about how to use a value. However, default is typically used to express that if a value is missing, then the value is semantically the same as if the value was present with the default value. The value of default should validate against the schema in which it resides, but that isn’t required.

The Value.Create function falls into the "Non-Validation Tools" category, where this function can be used outside of validation pipelines to initialize a value. The Value.Create function will initialize a value with the `default` value if present, and will mandate that the `default` value is specified in scenarios where TypeBox cannot initialize the value from the type alone...these cases include `format`, `regex` and some composite types such as illogical intersection.

If using Value.Create and you receive an error trying to create the value, the recommendation is to ensure the type that is causing the error have a default value, and that the default value matches the anticipated type.

---

> Are these some default formats or we have to register our own formats using FormatRegistry? For example I'm assuming that just setting the format to 'email' would assume that the input now should just be an email without me adding a format explicitly:

TypeBox does not implement any formats by default (not even the ones listed for the `format` property). So any custom formats you see in Elysia would be implemented by Elysia (and likely by way of the `FormatRegistry`). You are free to add additional formats on top of the ones implemented by Elysia if you need (by registering them in the FormatRegistry, Elysia should be able to use them).

There has been some consideration towards TypeBox implementing the default `formats` listed on the string options, however implementation would require full compliance with the RFC's for each format (And I haven't found the time to implement them all). However I have been tracking the current "best" implementations at the following URL (which is derived from the derived from the [ajv-formats](https://github.com/ajv-validator/ajv-formats) package)

https://github.com/sinclairzx81/typebox/tree/master/examples/formats

Until a complete set of validators can be established, TypeBox requires integrators to define their own formats. This would be inline with Elysia's implementation.

Hope this helps
S

**Comment by bhuynhdev** [comment]
Maybe I can chime in a bit here. In the specific context of Elysiajs, I found that Elysia does seem to use `Value.create` to help generate an error message whenever a VALIDATION error occurs: https://github.com/elysiajs/elysia/blob/78671b1827ecb40add10c6a0fab28ce4187f28ef/src/error.ts#L125

So even though it's just for validation, Elysia is still needing to call `Value.create`, thus leading to this error message.
